### PR TITLE
Retry ILM force merge step on shard failures

### LINKED ISSUE #73142: Retry ILM force merge step on shard failures
Today ILM's `forcemerge` step treats any non-exception response as success:

https://github.com/elastic/elasticsearch/blob/09d1b97b5f566481d8395fa04618e6ee1da21d28/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/ForceMergeStep.java#L44-L46

This is a very weak measure of success, it means that the coordination of the request was successful and not that any merging actually took place. More specifically it ignores `response.getFailedShards()` which is nonzero if some shards could not be merged, perhaps because they were initialising at the time the step executes.

If we ignore an initialising replica here then it may be promoted to primary later, which means that other shard copies and snapshots will continue to contain multiple segments as if the `forcemerge` step did not exist. In particular we prefer searchable snapshots to use single-segment shards for performance reasons but we cannot rely on ILM to generate such shards.

Also relates https://github.com/elastic/elasticsearch/issues/17094#issuecomment-838043706: if we were to permit users to cancel ongoing force-merge tasks then ILM should not carry on regardless.

I think we should retry this step until no shards report failures.

Note that today "no shard failures" is not a watertight guarantee that all shard copies will end up force-merged, because we do not apply the force-merge action to relocation targets or account for shard movements that start while the force-merge is ongoing. We can fix that too, and I would prefer to do that by reporting these unmerged shards as failures once ILM respects that.

### PR DESCRIPTION
**Pull request by andreidan**
This makes the forcemerge step inspect the response of the forcemerge
operation and retry if there are any failed shards.

Closes #73142

### CODE REVIEW

**Comment by andreidan** [comment]
@elasticmachine update branch

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/ForceMergeStep.java [hunk 1]
```diff
@@ -39,10 +48,29 @@ public int getMaxNumSegments() {
...
+                        String errorMessage =
+                            String.format(Locale.ROOT, "index [%s] in policy [%s] encountered failures [%s] on step [%s]",
+                                indexName, policyName,
+                                failures == null ? "n/a" : Strings.collectionToDelimitedString(Arrays.stream(failures)
+                                    .map(Strings::toString)
+                                    .collect(Collectors.toList()), ","),
+                                NAME);
+                        logger.warn(errorMessage);
+                        // let's report it as a failure and retry
+                        listener.onFailure(new IllegalStateException(errorMessage));
```
Comment (dakrone) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/ForceMergeStep.java:71]: I'm trying to think if there is a better exception type to use here.. maybe we should use `ElasticsearchException` with your current message and the first `DefaultShardOperationFailedException` passed as the `cause`.

What do you think?

**Review by dakrone — APPROVED** [review]
This LGTM, I did leave a comment about the type of exception though, thanks for fixing this!

**Comment by andreidan** [comment]
@elasticmachine update branch

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/ForceMergeStep.java (hunk 1 above)
Comment (andreidan) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/ForceMergeStep.java]: I changed it to use `ElasticsearchException` but very misleadingly `DefaultShardOperationFailedException` is **not** an `Exception` so we'll have to carry on with a null `cause`

Comment (dakrone) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/ForceMergeStep.java]: Okay, that's unfortunate naming indeed, but thanks for the explanation :)

**Pull request merged**